gio-subclass/src/auto/action_map.rs

```diff
@@ -0,0 +1,144 @@
+// from gir-files (https://github.com/gtk-rs/gir-files @ 47c69e6)
+// DO NOT EDIT
+
+#[allow(unused_imports)]
+use libc::{c_int, c_char, c_uchar, c_float, c_uint, c_double,
+    c_short, c_ushort, c_long, c_ulong,
+    c_void, size_t, ssize_t, intptr_t, uintptr_t, time_t, FILE};
+
+#[allow(unused_imports)]
+use glib_ffi::{gboolean, gconstpointer, gpointer, GType};
+
+use gio;
+use gio_ffi;
+use glib;
+use glib::object::IsA;
+use glib::translate::*;
+use glib_ffi;
+use gobject_ffi;
+use std::mem;
+use std::ptr;
+
+use gobject_subclass::anyimpl::*;
+use gobject_subclass::object::*;
+
+
+pub trait ActionMapImpl: AnyImpl + 'static {
+
+    fn add_action(&self, action_map: &gio::ActionMap, action: &gio::Action);
+
+    fn lookup_action(&self, action_map: &gio::ActionMap, action_name: &str) -> Option<gio::Action>;
+
+    fn remove_action(&self, action_map: &gio::ActionMap, action_name: &str);
+
+}
+
+any_impl!(ActionMapImpl);
+
+pub trait ActionMapImplStatic<T: ObjectType>: 'static {
+    fn get_impl<'a>(&self, imp: &'a T::ImplType) -> &'a ActionMapImpl;
+}
+
+struct ActionMapStatic<T: ObjectType>{
+    imp_static: *const ActionMapImplStatic<T>
+}
+
+
+unsafe extern "C" fn action_map_add_action<T: ObjectType>
+(gptr: *mut gio_ffi::GActionMap, action: *mut gio_ffi::GAction)
+{
+    floating_reference_guard!(gptr);
+    let klass = &**(gptr as *const *const ClassStruct<T>);
+    let interface_static = klass.get_interface_static(gio_ffi::g_action_map_get_type())
+                                     as *const ActionMapStatic<T>;
+    let instance = &*(gptr as *const T::InstanceStructType);
+    let imp = instance.get_impl();
+    let imp = (*(*interface_static).imp_static).get_impl(imp);
+    let wrap = from_glib_borrow(gptr);
+    imp.add_action(&wrap, &from_glib_none(action));
+}
+
+unsafe extern "C" fn action_map_lookup_action<T: ObjectType>
+(gptr: *mut gio_ffi::GActionMap, action_name: *const c_char) -> *mut gio_ffi::GAction
+{
+    floating_reference_guard!(gptr);
+    let klass = &**(gptr as *const *const ClassStruct<T>);
+    let interface_static = klass.get_interface_static(gio_ffi::g_action_map_get_type())
+                                     as *const ActionMapStatic<T>;
+    let instance = &*(gptr as *const T::InstanceStructType);
+    let imp = instance.get_impl();
+    let imp = (*(*interface_static).imp_static).get_impl(imp);
+    let wrap = from_glib_borrow(gptr);
+    let rs_ret = imp.lookup_action(&wrap, &String::from_glib_none(action_name));
+    
+    match rs_ret {
+        Some(t_ret) => {
+        let ret = t_ret.to_glib_full();
+        unsafe extern "C" fn destroy_ret(p: glib_ffi::gpointer){
+            let _: gio::Action = gio::Action::from_glib_full(p as *mut gio_ffi::GAction);
+        };
+        gobject_ffi::g_object_set_qdata_full(gptr as *mut gobject_ffi::GObject,
+            glib_ffi::g_quark_from_string("rs_action_map_lookup_action_ret".to_glib_none().0),
```

Comment:
I think this is not correct here. The caller should be able to grab all actions from the object (different names), but with your code now the action would potentially become invalid at the next call for a differently named action.

```rust
let action1 = map.lookup_action("action1");
let action2 = map.lookup_action("action2");
// at this point action1 is invalid but shouldn't
```

Reply:
Indeed. Any ideas how this can be solved within the generator? For the return value I could marshall the input value(s), but that's never going to be the silver bullet anyway...

Reply:
Do you have any idea how this is solved in the python bindings? I would assume they would've run into a similar situation at some point?

Reply:
> Indeed. Any ideas how this can be solved within the generator? For the return value I could marshall the input value(s), but that's never going to be the silver bullet anyway...

As this is a map, the actions should be valid until they are removed.

> Do you have any idea how this is solved in the python bindings? I would assume they would've run into a similar situation at some point?

Good question, I don't know. That's definitely worth checking :)

Reply:
Just for clarity; would this work with my current implementation? In any language, not just Rust.

```
let action1a = map.lookup_action("action1");
let action1b = map.lookup_action("action1");
// is action1a still valid here?
```

Reply:
Yes, it basically should behave like a map/hashtable kind of data structure. You add things, then you can retrieve them (and references are valid) until you remove them again

Reply:
I should rephrase my question. What I meant is: does calling ```to_glib``` for a rust structure always result in the same pointer?

Reply:
Unfortunately no

Reply:
So even like it currently stands, it would not even work when retrieving the same action twice. 
How bad would it be to just assign the quark data a totally random suffix?
This type of problem will undoubtedly also be present in other objects/interfaces, so I'm trying to come up with something that would work globally.

Reply:
> So even like it currently stands, it would not even work when retrieving the same action twice.

Well, in this specific case (with refcounted types) you always get the same pointer, so that's not really a problem :) Just in general that's not the case

> How bad would it be to just assign the quark data a totally random suffix?

Quite bad because you would leak everything until the object is destroyed. Removing from the map would not cause things to get destroyed.

> so I'm trying to come up with something that would work globally.

I would first take a look at how it's done in python, maybe that gives some ideas.

Reply:
I believe this is where I should look, but it's a lot to go through without prior knowledge.
https://gitlab.gnome.org/GNOME/pygobject/blob/master/gi/pygi-object.c

Reply:
From what I can tell, in PyGObject, they store the C pointer alongside the Python type in a global cache. I haven't found where they remove the item from the cache, though. I suppose that in python they could just hook into the refcounting bits of the interpreter and run destroy whenever it reaches zero?

Reply:
That's also not sufficient, you need to keep the thing alive as long as there is a reference anywhere (even outside Python). It probably uses the GObject toggleref feature for getting notified when that happens, but not sure how that helps in this specific case. Togglerefs are for the specific case where you need to keep data alive that is stored outside the actual GObject, e.g. the Python wrapper object around the GObject. We don't have such a problem.

You need to return something and at least keep it alive long enough for the caller to ref it again, so you need to store a strong reference somewhere at least.

Reply:
Since, according to the docs, ```ActionMap``` takes a ref itself, I would think that calling the ```destroy``` function for the qdata when removing the action from the map would be fine. I do feel, however, that that's a bit redundant. The rust implementation itself will also hold on to the ```Action```s. But I can't think of any other solution, TBH.

Reply:
From what I can see, Python just puts the burden of not causing crashes on the developer.

So, to move forward here, what I would suggest for now is to simply mark this interface as `unsafe`. That's basically the Rust way of saying "it's your job from preventing this from crashing". And keep the old way of just storing the very last return value in qdata to ensure it stays alive at least for a bit.

What do you think?

Reply:
@vhdirk ?

Reply:
No worries :)

I also noticed something else we probably want. Default vfunc implementations for interfaces, not sure if this is already relevant for any interfaces that you added.

See e.g. https://github.com/sdroege/gst-plugin-rs/commit/d1f90d0feaf29cbc7f27ce65858b4b445ac48b71#diff-e19d43e6ae8224d1a5d7c511c1fcef33R21

Reply:
Seems useful. The priority of this has shifted back a bit. I've been working om my application, taking inspiration from Hammond and Fractal. Currently, using gtk-subclass still adds so much boilerplate that it still a hassle to actually use it... (aside from cellrenderer, that is)

Reply:
Yes, I agree. It only really makes sense where there is no other option currently